File: BibliotecaUFABC/app/src/main/java/com/nintersoft/bibliotecaufabc/webviewclients/ReservationWebClient.java
package com.nintersoft.bibliotecaufabc.webviewclients;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.Build;
import android.webkit.ValueCallback;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.nintersoft.bibliotecaufabc.ReservationActivity;
import com.nintersoft.bibliotecaufabc.utilities.GlobalConstants;
import com.nintersoft.bibliotecaufabc.utilities.GlobalFunctions;

import androidx.annotation.RequiresApi;

import org.json.JSONException;
import org.json.JSONObject;

public class ReservationWebClient extends WebViewClient {
    private int reservation_page_finished;
    private int confirmation_page_finished;

    private Context mContext;

    public ReservationWebClient(Context context){
        super();
        this.mContext = context;
        reservation_page_finished = 0;
        confirmation_page_finished = 0;
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    @Override
    public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
        return false;
    }

    @Override
    public boolean shouldOverrideUrlLoading(WebView view, String url) {
        return false;
    }

    @Override
    public void onPageStarted(WebView view, String url, Bitmap favicon) {
        super.onPageStarted(view, url, favicon);
    }

    @Override
    public void onPageFinished(WebView view, String url) {
        super.onPageFinished(view, url);

        if (url.contains(GlobalConstants.URL_LIBRARY_CANCEL_RESERVATION)){
            if (confirmation_page_finished < 1){
                confirmation_page_finished++;
                return;
            }

            String script = String.format("%1$s \ngetCancellationMessage();",
                    GlobalFunctions.getScriptFromAssets(mContext, "javascript/reservation_scraper.js"));
            view.evaluateJavascript(script, new ValueCallback<String>() {
                @Override
                public void onReceiveValue(final String value) {
                    ((ReservationActivity)mContext).runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            ((ReservationActivity)mContext).setupInterface(false);
                            ((ReservationActivity)mContext).showCancellationMessage(value);
                        }
                    });
                }
            });
            reservation_page_finished = 0;
        }
        else if (url.contains(GlobalConstants.URL_LIBRARY_RESERVATION)){
            if (reservation_page_finished < 1){
                reservation_page_finished++;
                return;
            }

            String script = String.format("%1$s \ngetReservations();",
                    GlobalFunctions.getScriptFromAssets(mContext, "javascript/reservation_scraper.js"));
            view.evaluateJavascript(script, new ValueCallback<String>() {
                @Override
                public void onReceiveValue(final String value) {
                    ((ReservationActivity)mContext).runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            try{
                                JSONObject result = new JSONObject(value);
                                if (result.getBoolean("hasErrorDiv"))
                                    ((ReservationActivity)mContext).setUserNameNoReservation(result.getString("username"));
                                else {
                                    ((ReservationActivity)mContext).setReservationBooks(result.getJSONArray("reservationBooks").toString());
                                    ((ReservationActivity)mContext).setupInterface(true);
                                }
                            } catch (JSONException e){
                                ((ReservationActivity)mContext).runOnUiThread(new Runnable() {
                                    @Override
                                    public void run() {
                                        ((ReservationActivity)mContext).setUserNameNoReservation("???");
                                    }
                                });
                            }
                        }
                    });
                }
            });
            confirmation_page_finished = 0;
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    @Override
    public void onReceivedError(WebView view, WebResourceRequest request, final WebResourceError error) {
        super.onReceivedError(view, request, error);

        ((ReservationActivity)mContext).runOnUiThread(new Runnable() {
            @Override
            public void run() {
                ((ReservationActivity)mContext).setErrorForm(error.getDescription().toString());
            }
        });
    }

    @Override
    public void onReceivedError(WebView view, int errorCode, final String description, String failingUrl) {
        super.onReceivedError(view, errorCode, description, failingUrl);

        ((ReservationActivity)mContext).runOnUiThread(new Runnable() {
            @Override
            public void run() {
                ((ReservationActivity)mContext).setErrorForm(description);
            }
        });
    }
}

File: BibliotecaUFABC/app/src/main/java/com/nintersoft/bibliotecaufabc/webclient/ReservationWebClient.kt
package com.nintersoft.bibliotecaufabc.webclient

import android.os.Build
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.annotation.RequiresApi
import com.nintersoft.bibliotecaufabc.R
import com.nintersoft.bibliotecaufabc.global.Constants
import com.nintersoft.bibliotecaufabc.global.Functions
import com.nintersoft.bibliotecaufabc.model.AppContext
import com.nintersoft.bibliotecaufabc.ui.reservations.ReservationsViewModel
import org.json.JSONException
import org.json.JSONObject

class ReservationWebClient(private val viewModel : ReservationsViewModel)
    : WebViewClient() {

    private val context = AppContext.context
    private var reservationPageFinished = 0
    private var confirmationPageFinished = 0

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
        return false
    }

    override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean { return false }

    override fun onPageFinished(view: WebView?, url: String?) {
        super.onPageFinished(view, url)
        if (url.isNullOrEmpty()) return

        if (url.contains(Constants.URL_LIBRARY_CANCEL_RESERVATION)){
            if (confirmationPageFinished < 1){
                confirmationPageFinished++
                return
            }

            val script = Functions.scriptFromAssets("js/reservation_scp.js") + "\ngetCancellationMessage();"
            view?.evaluateJavascript(script, ({
                viewModel.setCancellationMessage(it)
                viewModel.requestReload(true)
            }))
            reservationPageFinished = 0
        }
        else if (url == Constants.URL_LIBRARY_RESERVATION){
            if (reservationPageFinished < 1){
                reservationPageFinished++
                return
            }

            val script = Functions.scriptFromAssets("js/reservation_scp.js") + "\ngetReservations();"
            view?.evaluateJavascript(script, ({
                try{
                    with(JSONObject(it)){
                        if (getBoolean("hasErrorDiv"))
                            viewModel.setUserName(getString("username"))
                        else {
                            viewModel.setUserName(null)
                            viewModel.setReservedBooks(getJSONArray("reservationBooks"))
                        }
                    }
                }
                catch (_ : JSONException){ viewModel.setUserName(context?.
                    getString(R.string.on_error_def_username)) }
                finally { reservationPageFinished = 0 }
            }))
            confirmationPageFinished = 0
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    override fun onReceivedError(view: WebView?, request: WebResourceRequest?,
                                 error: WebResourceError?) {
        super.onReceivedError(view, request, error)
        reservationPageFinished = 0
        confirmationPageFinished = 0
        viewModel.setLoadError(true)
    }

    @Suppress("DEPRECATION")
    override fun onReceivedError(view: WebView?, errorCode: Int, description: String?,
        failingUrl: String?) {
        super.onReceivedError(view, errorCode, description, failingUrl)
        reservationPageFinished = 0
        confirmationPageFinished = 0
        viewModel.setLoadError(true)
    }

}